examples/vision/nerf.py

```diff
@@ -0,0 +1,810 @@
+ReLU activation and 128 channels) that output the view-dependent RGB
+color."
+
+Here we have gone for a minimal implementation and have used `64`
+Dense units instead of `256` as that mentioned in the paper.
+"""
+
+
+def get_nerf_model(num_layers, num_pos):
+    # Initialize the input layer.
+    inputs = keras.Input(shape=(num_pos, 2 * 3 * POS_ENCODE_DIMS + 3))
+    x = inputs
+    for i in range(num_layers):
+        # Define the Dense layers
+        x = layers.Dense(units=64, activation="relu")(x)
+        if i % 4 == 0 and i > 0:
+            # Inject residual connection.
+            x = layers.concatenate([x, inputs], axis=-1)
+
+    # Process the output.
+    outputs = layers.Dense(units=4)(x)
+
+    # Create and return the model.
+    return keras.Model(inputs=inputs, outputs=outputs)
+
+
+def render_rgb_depth(model, rays_flat, t_vals, rand=True, train=True):
+    """
+    Generates the rgb image and depth map from the prediction of the
+    model.
+    Args:
+        - model: The MLP model that is trained to predict the rgb and
+            volume density of the volumetric scene.
+        - rays_flat: The flattened rays that serves as the input to
+            the NeRF model.
+        - t_vals: The sample points for the rays.
+        - rand: Choice to randomise the sampling strategy.
+        - train: Whether the model is in training or testing phase.
+    """
+    # Get the predictions from the nerf model and reshape it.
+    if train:
+        predictions = model(rays_flat)
+    else:
+        predictions = model.predict(rays_flat)
+    predictions = tf.reshape(predictions, shape=(BATCH_SIZE, H, W, NUM_SAMPLES, 4))
+
+    # Slice the predictions into rgb and sigma.
+    rgb = tf.sigmoid(predictions[..., :-1])
+    sigma_a = tf.nn.relu(predictions[..., -1])
+
+    # Get the distance of adjacent intervals.
+    delta = t_vals[..., 1:] - t_vals[..., :-1]
+    # delta shape = (num_samples)
+    if rand:
+        delta = tf.concat(
+            [delta, tf.broadcast_to([1e10], shape=(BATCH_SIZE, H, W, 1))], axis=-1
+        )
+        alpha = 1.0 - tf.exp(-sigma_a * delta)
+    else:
+        delta = tf.concat(
+            [delta, tf.broadcast_to([1e10], shape=(BATCH_SIZE, 1))], axis=-1
+        )
+        alpha = 1.0 - tf.exp(-sigma_a * delta[:, None, None, :])
+
+    # Get transmittance.
+    exp_term = 1.0 - alpha
+    epsilon = 1e-10
+    transmittance = tf.math.cumprod(exp_term + epsilon, axis=-1, exclusive=True)
+    weights = alpha * transmittance
+    rgb = tf.reduce_sum(weights[..., None] * rgb, axis=-2)
+
+    if rand:
+        depth_map = tf.reduce_sum(weights * t_vals, axis=-1)
+    else:
+        depth_map = tf.reduce_sum(weights * t_vals[:, None, None], axis=-1)
+    return (rgb, depth_map)
+
+
+"""
+## Training
+
+The training step is gathered in a `keras.Model` class so that we can
+make use of the `model.fit` call.
+"""
+
+loss_tracker = keras.metrics.Mean(name="loss")
+psnr_metric = keras.metrics.Mean(name="psnr")
```

Comment:
Prefer creating them in `__init__()` or `compile()` and attaching them to the model, to make the model self-contained

Reply:
Insightful. Sorry for not understanding this in the first pass.